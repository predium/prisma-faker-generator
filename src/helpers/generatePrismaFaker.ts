import { DMMF } from '@prisma/generator-helper';
import _ from 'lodash';

class Model {
  readonly scalars: DMMF.Field[];
  readonly foreignKeys: string[];
  readonly scalarsRequiredForCreation: DMMF.Field[];
  readonly objectsRequiredForCreation: DMMF.Field[];
  readonly name: string;

  constructor(dmmf_model: DMMF.Model) {
    this.name = dmmf_model.name;
    this.scalars = dmmf_model.fields.filter((field) => field.kind === 'scalar');
    this.foreignKeys = dmmf_model.fields.flatMap((field) => field.relationFromFields || []);

    this.scalarsRequiredForCreation = this.scalars.filter(
      (field) => field.isRequired && !field.hasDefaultValue && !this.foreignKeys.includes(field.name),
    );

    this.objectsRequiredForCreation = dmmf_model.fields.filter(
      (field) => field.kind === 'object' && !field.isList && field.isRequired,
    );
  }
}

class PrismaFakerGenerator {
  models: Model[];

  readonly scalarMap: Record<string, string> = {
    Int: 'faker.number.int({ max: 1000 })',
    BigInt: 'faker.number.bigInt()',
    Float: 'faker.number.float()',
    String: 'faker.string.sample()',
    DateTime: 'faker.date.anytime ()',
    Decimal: 'new Prisma.Decimal(faker.number.float())',
    Boolean: 'faker.datatype.boolean()',
    Json: '{}',
  };

  constructor(models: DMMF.Model[]) {
    this.models = models.map((dmmfModel) => new Model(dmmfModel));
  }

  generate(): string {
    return [this.header, this.generateBody(), this.footer].join('\n\n');
  }

  readonly header = `import { faker } from '@faker-js/faker';
import * as prisma from '@prisma/client';
import { Prisma } from '@prisma/client';

export default class prisma_faker {`;

  private generateBody(): string {
    return `${this.generateDatabaseFaker()}

${this.generateTypescriptFaker()}`;
  }

  readonly footer = `}`;

  //General////////////////////////////////////////////////////////////////////

  generateScalarFaker(field: DMMF.Field): string {
    if (field.isList) {
      return '[]';
    }
    return `${this.scalarMap[field.type]}`;
  }

  generateScalarFieldFaker(field: DMMF.Field): string {
    return `    ${field.name}: ${this.generateScalarFaker(field)},`;
  }

  //Typescript/////////////////////////////////////////////////////////////////
  private generateTypescriptFaker(): string {
    const typescriptModelFakers = this.models.map((model) => this.generateTypescriptModelFaker(model)).join('\n\n');

    const fakersExport = this.models.map((model) => `    ${model.name}: prisma_faker.${model.name}_faker,`);

    const typescriptFakersCollection = `
  static readonly typescript = {
${fakersExport.join('\n')}
  };`;
    return [typescriptModelFakers, typescriptFakersCollection].join(`\n`);
  }

  private generateTypescriptModelFaker(model: Model): string {
    const signature = `  static readonly ${model.name}_faker = (partial_${model.name}?: Partial<prisma.${model.name}>): prisma.${model.name} => ({`;

    const fields = model.scalars.map((field) => this.generateScalarFieldFaker(field));

    const footer = `    ...partial_${model.name},
  });`;
    return [signature, ...fields, footer].join('\n');
  }

  //Database///////////////////////////////////////////////////////////////////

  generateDatabaseFaker() {
    const dbModelFakers = this.models.map((model) => this.generateDatabaseModelFaker(model)).join(`\n\n`);

    const fakersExport = this.models.map((model) => `    ${model.name}: prisma_faker.${model.name}_faker_db,`);

    const dbFakersCollection = `
  static readonly db = {
${fakersExport.join('\n')}
  };`;
    return [dbModelFakers, dbFakersCollection].join(`\n`);
  }

  generateDatabaseModelFaker(model: Model): any {
    const signature = `  static readonly ${model.name}_faker_db = (partial_${model.name}?: Partial<Prisma.${model.name}CreateInput>): Prisma.${model.name}CreateInput => ({`;

    const requiredScalars = model.scalarsRequiredForCreation.map((field) => this.generateScalarFieldFaker(field));

    const requiredObjects = model.objectsRequiredForCreation.map(
      (field) => `    ${field.name}: { create: prisma_faker.${field.type}_faker_db()},`,
    );

    const footer = `    ...partial_${model.name},
  });`;
    return [signature, ...requiredObjects, ...requiredScalars, footer].join(`\n`);
  }
}

export function generatePrismaFaker(models: DMMF.Model[]): string {
  const generator = new PrismaFakerGenerator(models);
  return generator.generate();
}
