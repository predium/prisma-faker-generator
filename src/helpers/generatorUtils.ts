import { GeneratorOptions } from '@prisma/generator-helper';
import { logger } from '@prisma/internals';

export function getMapperOptions(config: GeneratorOptions['generator']['config']) {
  const mapper = config.mapper;
  if (mapper) {
    if (!(mapper instanceof Array)) {
      throw new Error(`Mapper is not an array`);
    }

    const parsedMapper: { regex: string; faker: string }[] = mapper.map((m) => {
      const split = m.split('->');
      if (split.length !== 2) {
        throw new Error(`Mapper is not in the correct format. Could not parse ${m}`);
      }
      const [regex, faker] = split;
      return { regex, faker };
    });

    logger.info(`Mapper found. Parsed ${parsedMapper.length} mappings`);
    return parsedMapper;
  }
}
