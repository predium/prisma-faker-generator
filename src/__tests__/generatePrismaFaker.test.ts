import { getConfig, getDMMF } from '@prisma/internals';
import { generatePrismaFaker } from '../helpers/generatePrismaFaker';
import { getMapperOptions } from '../helpers/generatorUtils';

it('should work on a simple model', async () => {
  const datamodel = `
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  email     String   @unique
  name      String?
  hashedPassword Bytes?
  posts     Post[]
}

model Post {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  published Boolean  @default(false)
  title     String   @db.VarChar(255)
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
}
`;

  const dmmf = await getDMMF({ datamodel });
  const result = generatePrismaFaker(dmmf.datamodel.models);

  expect(result).toMatchSnapshot();
});

describe('Mapper', () => {
  it('should work when the mapper is missing', async () => {
    const datamodel = `
datasource db {
  provider = "postgresql"
  url      = "postgresql://localhost:5432"
}

generator faker {
  provider = "prisma-faker-generator"
  output   = "../src/utils/testing"
  mapper   = ["Book.id->faker.number.int()", ".*\.id->faker.string.uuid()"]
}

model User {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  email     String   @unique
  name      String?
  posts     Post[]
}

model Post {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  published Boolean  @default(false)
  title     String   @db.VarChar(255)
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
}
`;

    const dmmf = await getDMMF({ datamodel });
    const configs = await getConfig({ datamodel });
    const result = generatePrismaFaker(dmmf.datamodel.models, getMapperOptions(configs.generators[0].config));

    expect(result).toMatchSnapshot();
  });

  it('should generate custom map fields', async () => {
    const datamodel = `
generator faker {
  provider = "prisma-faker-generator"
  output   = "../src/utils/testing"
  mapper   = ["Book.id->faker.number.int()", ".*\.id->faker.string.uuid()"]
}

datasource db {
  provider = "postgresql"
  url      = "postgresql://localhost:5432"
}

model Book {
    id                    String                      @unique
}

model User {
    id                    String                      @db.Uuid @unique
}`;

    const dmmf = await getDMMF({ datamodel });
    const configs = await getConfig({ datamodel });
    const mapper = getMapperOptions(configs.generators[0].config);
    const result = generatePrismaFaker(dmmf.datamodel.models, mapper);

    expect(result).toMatchSnapshot();
  });
});
