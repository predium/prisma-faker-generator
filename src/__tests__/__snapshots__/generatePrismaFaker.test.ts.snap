// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mapper should generate custom map fields 1`] = `
"import { faker } from '@faker-js/faker';
import * as prisma from '@prisma/client';
import { Prisma } from '@prisma/client';

export default class prisma_faker {

  static readonly Book_faker_db = (partial_Book?: Partial<Prisma.BookCreateInput>): Prisma.BookCreateInput => ({
    id: faker.number.int(),
    ...partial_Book,
  });

  static readonly User_faker_db = (partial_User?: Partial<Prisma.UserCreateInput>): Prisma.UserCreateInput => ({
    id: faker.string.uuid(),
    ...partial_User,
  });

  static readonly db = {
    Book: prisma_faker.Book_faker_db,
    User: prisma_faker.User_faker_db,
  };

  static readonly Book_faker = (partial_Book?: Partial<prisma.Book>): prisma.Book => ({
    id: faker.number.int(),
    ...partial_Book,
  });

  static readonly User_faker = (partial_User?: Partial<prisma.User>): prisma.User => ({
    id: faker.string.uuid(),
    ...partial_User,
  });

  static readonly typescript = {
    Book: prisma_faker.Book_faker,
    User: prisma_faker.User_faker,
  };

}"
`;

exports[`Mapper should work when the mapper is missing 1`] = `
"import { faker } from '@faker-js/faker';
import * as prisma from '@prisma/client';
import { Prisma } from '@prisma/client';

export default class prisma_faker {

  static readonly User_faker_db = (partial_User?: Partial<Prisma.UserCreateInput>): Prisma.UserCreateInput => ({
    email: faker.string.sample(),
    ...partial_User,
  });

  static readonly Post_faker_db = (partial_Post?: Partial<Prisma.PostCreateInput>): Prisma.PostCreateInput => ({
    updatedAt: faker.date.anytime(),
    title: faker.string.sample(),
    ...partial_Post,
  });

  static readonly db = {
    User: prisma_faker.User_faker_db,
    Post: prisma_faker.Post_faker_db,
  };

  static readonly User_faker = (partial_User?: Partial<prisma.User>): prisma.User => ({
    id: faker.string.uuid(),
    createdAt: faker.date.anytime(),
    email: faker.string.sample(),
    name: faker.string.sample(),
    ...partial_User,
  });

  static readonly Post_faker = (partial_Post?: Partial<prisma.Post>): prisma.Post => ({
    id: faker.string.uuid(),
    createdAt: faker.date.anytime(),
    updatedAt: faker.date.anytime(),
    published: faker.datatype.boolean(),
    title: faker.string.sample(),
    authorId: faker.number.int({ max: 1000 }),
    ...partial_Post,
  });

  static readonly typescript = {
    User: prisma_faker.User_faker,
    Post: prisma_faker.Post_faker,
  };

}"
`;

exports[`should work on a simple model 1`] = `
"import { faker } from '@faker-js/faker';
import * as prisma from '@prisma/client';
import { Prisma } from '@prisma/client';

export default class prisma_faker {

  static readonly User_faker_db = (partial_User?: Partial<Prisma.UserCreateInput>): Prisma.UserCreateInput => ({
    email: faker.string.sample(),
    ...partial_User,
  });

  static readonly Post_faker_db = (partial_Post?: Partial<Prisma.PostCreateInput>): Prisma.PostCreateInput => ({
    updatedAt: faker.date.anytime(),
    title: faker.string.sample(),
    ...partial_Post,
  });

  static readonly db = {
    User: prisma_faker.User_faker_db,
    Post: prisma_faker.Post_faker_db,
  };

  static readonly User_faker = (partial_User?: Partial<prisma.User>): prisma.User => ({
    id: faker.number.int({ max: 1000 }),
    createdAt: faker.date.anytime(),
    email: faker.string.sample(),
    name: faker.string.sample(),
    hashedPassword: Buffer.from(faker.string.sample()),
    ...partial_User,
  });

  static readonly Post_faker = (partial_Post?: Partial<prisma.Post>): prisma.Post => ({
    id: faker.number.int({ max: 1000 }),
    createdAt: faker.date.anytime(),
    updatedAt: faker.date.anytime(),
    published: faker.datatype.boolean(),
    title: faker.string.sample(),
    authorId: faker.number.int({ max: 1000 }),
    ...partial_Post,
  });

  static readonly typescript = {
    User: prisma_faker.User_faker,
    Post: prisma_faker.Post_faker,
  };

}"
`;
